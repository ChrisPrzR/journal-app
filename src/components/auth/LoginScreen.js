import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';

import { removeError, setError } from '../../actions/ui';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';

export const LoginScreen = () => {


    const dispatch = useDispatch();
    const {msgError, loading} = useSelector( state => state.ui );

    const [ formValues, handleInputChange ] = useForm({
        email: '',
        password: ''
    });

    const {email, password} = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        isLoginValid() && dispatch( startLoginEmailPassword(email, password) )
    }

    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin())
    }

    const isLoginValid = () => {
        if(!validator.isEmail( email )){
            dispatch(setError('Must be a valid email'))
            return false
        }else if(password.length < 6 ){
            dispatch(setError('Password must be 6 characters or more'))
            return false
        }
        dispatch(removeError())
        return true
    }

    return (
        <div>
            <h3 className="auth__title">Login</h3>
            <form onSubmit={ handleLogin }>
            {
                msgError &&
                (<div className="auth__alert-error">{`${msgError}`}</div>)
                
            }
                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={email}
                    onChange={ handleInputChange }
                />
                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={password}
                    onChange={ handleInputChange }
                />

                {
                    loading ? 
                    <div>
                        <i className="fas fa-spinner btn btn-primary"></i>
                    </div>
                    :
                    <button
                    type="submit"
                    className="btn btn-primary btn-block"
                    >
                    Login
                    </button>
                }
                

                <hr/>
                <div className="auth__social-networks">
                    <p>Social Login</p>
                    <div 
                    className="google-btn"
                    onClick={handleGoogleLogin}
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link to="/auth/register" className="link">Create new account</Link>
            </form>
        </div>
    )
}
