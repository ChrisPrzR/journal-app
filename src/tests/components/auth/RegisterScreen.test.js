import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { MemoryRouter } from 'react-router-dom';

import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
};

let store = mockStore(initState);
// store.dispatch = jest.fn()

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <RegisterScreen />
        </MemoryRouter>
    </Provider>
)


describe('Testing <RegisterScreen/>', () => {
    
    test('should render ok', () => {
        
        expect(wrapper).toMatchSnapshot()

    });

    test('should dispatch the right action', () => {
        
        const emailField = wrapper.find('input[name="email"]');

        emailField.simulate('change', {
            target: {
                value: '',
                name: 'email'
            }
        });

        wrapper.find('form').simulate('submit',{
            preventDefault(){}
        });

        const actions = store.getActions()
        expect( actions[0] ).toEqual({
            type: types.uiSetError,
            payload: 'Email is not valid'
        });
    });

    test('should show alert box with error', () => {
        const initState = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'Invalid email'
            }
        };
        
        let store = mockStore(initState);
        // store.dispatch = jest.fn()
        
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );
        
        expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
    })    
})
