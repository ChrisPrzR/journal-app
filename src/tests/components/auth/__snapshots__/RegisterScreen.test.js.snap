// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <RegisterScreen/> should render ok 1`] = `
<div>
  <h3
    className="auth__title"
  >
    Register
  </h3>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Chris"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="correo@pruebas.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered? Login
    </a>
  </form>
</div>
`;
